### PR TITLE
feat: migrate advanced metadata (domain-settings)

### PR DESCRIPTION
**Pull request by domw30**
## 2. PR type
- feature

## 3. What is the old behaviour?
Domain Settings wizard/Advanced Metadata had not yet been migrated over to zOS.

## 4. What is the new behaviour?
Domain Settings wizard/Advanced Metadata migrated over to zOS

DEMO
https://www.loom.com/share/1b425c40063e42979ebddf917975911d

### CODE REVIEW

File: src/features/domain-settings/Steps/DetailsStep.module.scss [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+		&::-webkit-scrollbar {
+			width: 0.25rem;
+			background-color: none;
+		}
+
+		&::-webkit-scrollbar-thumb {
+			border-radius: 0.625rem;
+			background-color: rgba(60, 161, 255, 0.2);
+		}
```
Comment (colbr) [inline src/features/domain-settings/Steps/DetailsStep.module.scss:91]: Hmm maybe we need a utility component for this.

File: src/features/domain-settings/Steps/Footers/CompleteStepFooter.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	errorText,
+	confirmActionType,
+	onConfirmActionUpdate,
+	onStepUpdate,
+	onLockMetadataStatus,
+	onClose,
+}) => {
+	const onEdit = () => onStepUpdate(steps[0]);
+
+	return (
```
Comment (colbr) [inline src/features/domain-settings/Steps/Footers/CompleteStepFooter.tsx:31]: Same story as previous file, could break this JSX down into some subcomponents to make it easier to maintain.

File: src/features/domain-settings/hooks/useDomainSettingsForm.ts [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+			[domainId, details, provider.getSigner()],
+			{
+				onStart: () => handleTransactionStart(onLoadingHeader),
+				onProcessing: () => handleTransactionProcessing(onLoadingText),
+				onSuccess: () => handleTransactionSuccess(),
+				onError: (error: any) => handleTransactionError(error.message),
+				invalidationKeys: [['user', { account, domainId, details }]],
+			},
+		);
+	};
```
Comment (colbr) [inline src/features/domain-settings/hooks/useDomainSettingsForm.ts:206]: Could create an enum for `Save`, `Save & Lock`, `Lock`, combine these functions, and take an enum value as parameter. Then you'd just need to set `onLoadingHeader` and `onLoadingText` based on enum value. This would reduce the size of this file by quite a bit!

File: src/features/domain-settings/hooks/useDomainSettingsForm.ts [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+			domainId,
+			provider.getSigner(),
+		);
+
+		console.log('here');
+
+		console.log('DOMINA', domainMetadata);
+
+		return domainMetadata;
+	}, [domainId, provider]);
```
Comment (colbr) [inline src/features/domain-settings/hooks/useDomainSettingsForm.ts:119]: I wonder if instead of getting re-scraping lock status/metadata, we could invalidate the existing metadata query for the domain? That way it would populate the new domain lock status/metadata everywhere, rather than just the advanced metadata modal.

File: src/features/domain-settings/hooks/useDomainSettingsForm.ts [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+			isMintable,
+			gridViewByDefault,
+			customDomainHeader,
+			customDomainHeaderValue,
+		});
+	};
+
+	const onStepUpdate = (step: Step) => {
+		setStepId(step.id);
+	};
```
Comment (colbr) [inline src/features/domain-settings/hooks/useDomainSettingsForm.ts:89]: I think you may be able to skip these definitions, and just rename the `setState` functions in the hook return.

e.g.
```js
return {
    ...,
    onStepUpdate: setStepId,
    ...
}
```

File: src/features/ui/Switch/Switch.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import { FC } from 'react';
```
Comment (colbr) [inline src/features/ui/Switch/Switch.tsx:1]: We should add this one to `zUI`

**Review by colbr — COMMENTED** [review]
Looking nice all-in-all, but definitely some tidy up that would make this easier to maintain. Hopefully I left enough detail, but if not feel free to reach out 😊

File: src/features/domain-settings/DomainSettingsForm.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+		isTransactionLoading,
+		onStepUpdate,
+		onTitleUpdate,
+		onFormDetailsSubmit,
+		onLockMetadataStatus,
+		onSetAndLockMetadata,
+		onSetMetadata,
+		onConfirmActionUpdate,
+		onClose,
+	});
```
[the remark this answers is not in the document]
Comment (domw30) [reply src/features/domain-settings/DomainSettingsForm.tsx]: Nice yeah on second look the `useFormSteps` form content (i.e. `DetailsStep` & `ConfirmStep`) makes more sense in `useDomainSettingsForm`. Removed a lot of unnecessary code - `DomainSettingsForm` is now reduced to:

```
export const DomainSettingsForm: FC<DomainSettingsFormProps> = ({
	zna,
	onClose,
}) => {
	const {
		stepId,
		formHeader,
		formContent,
		isTransactionLoading,
		onStepUpdate,
	} = useDomainSettingsForm(zna, onClose);

	return (
		<Wizard.Container>
			<DomainSettingsHeader
				zna={zna}
				stepId={stepId}
				headerText={formHeader}
				isTransactionLoading={isTransactionLoading}
				onStepUpdate={onStepUpdate}
				onClose={onClose}
			/>
			<form>{formContent}</form>
		</Wizard.Container>
	);
};

```
So all of the form content and handlers are placed in `useDomainSettingsForm` and `DomainSettingsForm` is fed the content and whats required for the header. Thanks for the suggestions, makes it much easier to follow.

File: src/features/domain-settings/Steps/ConfirmStep.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+	onSetAndLockMetadata,
+	onSetMetadata,
+}) => {
+	const confirmationStepContent = getStepTextContent(confirmActionType);
+
+	const confirmationMessage = (
+		<FormTextContent textContent={confirmationStepContent.message} />
+	);
+
+	const onBack = () => onStepUpdate(steps[0]);
```
[the remark this answers is not in the document]
Comment (domw30) [reply src/features/domain-settings/Steps/ConfirmStep.tsx]: Done!

File: src/features/domain-settings/Steps/ConfirmStep.tsx [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+/************************
+ * getStepTextContent
+ ************************/
+
+type StepData = {
+	primaryButtonText: string;
+	message: string;
+};
+
+const getStepTextContent = (confirmActionType: string) => {
```
[the remark this answers is not in the document]
Comment (domw30) [reply src/features/domain-settings/Steps/ConfirmStep.tsx]: Good spot! Already had the enum in a types file. Now imported and using as the type. Thanks

File: src/features/domain-settings/Steps/ConfirmStep.tsx [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+			message:
+				'Your changes will be saved and the metadata will be locked. \nYou will be the only one who can unlock it in the future.',
+		};
+	} else {
+		stepData = {
+			primaryButtonText: 'Save Without Locking',
+			message:
+				'If you transfer ownership of the domain while metadata is unlocked, the new owner can edit the metadata and lock it. You may lose access forever.',
+		};
+	}
```
[the remark this answers is not in the document]
Comment (domw30) [reply src/features/domain-settings/Steps/ConfirmStep.tsx]: Sweet, done!

File: src/features/domain-settings/Steps/DetailsStep.tsx [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+			isBiddable,
+			gridViewByDefault,
+			customDomainHeader,
+			customDomainHeaderValue,
+		});
+	};
+
+	const isDisabled = metadataLockedStatus || stepId === FormStep.COMPLETE;
+
+	return !isLoadingSettingsData ? (
```
[the remark this answers is not in the document]
Comment (domw30) [reply src/features/domain-settings/Steps/DetailsStep.tsx]: Yeah very chunky. When a files this length it gives me the shivers. Pulled out some of the bulk into local subcomponents. Done! Let me know if you think this should be broken down any further. Thanks mate

File: src/features/domain-settings/Steps/DetailsStep.module.scss (hunk 1 above)
Comment (domw30) [reply src/features/domain-settings/Steps/DetailsStep.module.scss:91]: What would be the best way to create a utility component for use with `module.scss` files?

File: src/features/domain-settings/Steps/Footers/CompleteStepFooter.tsx (hunk 2 above)
Comment (domw30) [reply src/features/domain-settings/Steps/Footers/CompleteStepFooter.tsx:31]: Nice, done and will do for the other footer.

File: src/features/ui/Switch/Switch.tsx (hunk 6 above)
Comment (domw30) [reply src/features/ui/Switch/Switch.tsx:1]: For sure, there are a number of components within `zApp-NFTs` that need to be pulled over to `zUI`. If your happy to leave here for now, we can add a task to the back log to focus on actioning these?

File: src/features/domain-settings/hooks/useDomainSettingsForm.ts (hunk 3 above)
Comment (domw30) [reply src/features/domain-settings/hooks/useDomainSettingsForm.ts]: This file has expanded now with some of the changes, i.e. combining hooks! I've added a switch case to handle the loading text content as it was getting a bit messy within the transaction function itself. Let me know what you think of this one. Other files have reduced as a result of making these changes and only return one `onSubmitTransaction` rather than `onLockMetadata`, `onSetAndLockMetadata` and `onSetMetadata`.

File: src/features/domain-settings/hooks/useDomainSettingsForm.ts (hunk 5 above)
Comment (domw30) [reply src/features/domain-settings/hooks/useDomainSettingsForm.ts]: Removed these definitions.

File: src/features/domain-settings/hooks/useDomainSettingsForm.ts (hunk 4 above)
Comment (domw30) [reply src/features/domain-settings/hooks/useDomainSettingsForm.ts]: Yes that would be ideal! Do you have a suggestion of the best way to invalidate the existing metadata. I'm assuming this will then update what the latest data should be ?

Comment (domw30) [reply src/features/domain-settings/hooks/useDomainSettingsForm.ts]: As that's one of the issues currently when reaching the final step of the form, or closing the form once finished

Comment (domw30) [reply src/features/domain-settings/hooks/useDomainSettingsForm.ts]: Might be worth pulling this branch and giving it a run. Seems to go well locally for me, would be good to verify that though!

File: src/features/ui/Switch/Switch.tsx (hunk 6 above)
Comment (colbr) [reply src/features/ui/Switch/Switch.tsx:1]: Sounds good, create a backlog task for it and we can do it next sprint or whenever it comes up!

File: src/features/domain-settings/hooks/useDomainSettingsForm.ts (hunk 4 above)
Comment (colbr) [reply src/features/domain-settings/hooks/useDomainSettingsForm.ts]: I do! I did a [similar thing in the Staking zApp](https://github.com/zer0-os/zApp-Staking/blob/master/src/features/stake/useStake.ts), because I needed to refresh staking data after a successful stake/unstake. In short, you just need to invalidate the `react-query` keys you want to reset, and you do it directly on the query client.

Something like this:
```js
const queryClient = useQueryClient();

const onUpdateMetadata = () => {
    queryClient.invalidateQueries(['key ']);
}
```

There is probably better ways to do this, as it looks like `useQuery` also returns `refetch` and `remove` functions. Maybe have a dig into those, and if they don't yield any results, use what I chucked above. The only flaw with above is it's slightly verbose, and the invalidation key has to be identical to the query key.

File: src/features/domain-settings/Steps/ConfirmStep.tsx [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+			stepData = {
+				primaryButtonText: 'Save Without Locking',
+				message:
+					'If you transfer ownership of the domain while metadata is unlocked, the new owner can edit the metadata and lock it. You may lose access forever.',
+			};
+			break;
+	}
+
+	return stepData;
+};
```
Comment (colbr) [inline src/features/domain-settings/Steps/ConfirmStep.tsx:86]: nit: could shrink this down to a data structure rather than a function, e.g.

```js
const STEP_TEXT = [
    [ConfirmActionType.UNLOCK]: {
        primaryButtonText: 'Unlock Metadata',
        message: 'Unlocking metadata is a blockchain transaction that will cost gas. \nAdditional, optional, transactions are required to save changes and lock the metadata again.'
    },
    [ConfirmActionType.SAVE_AND_LOCK]: {
        primaryButtonText: 'Save & Lock',
        message: 'Your changes will be saved and the metadata will be locked. \nYou will be the only one who can unlock it in the future.'
    },
    ...
]

const { primaryButtonText, message } = STEP_TEXT[ConfirmActionType.UNLOCK]
```

Have a wee look at `NETWORK_CONFIGS` in `network.ts` for an example with type safety.

Using a function is the way to go when some additional calculations might be needed, however, this is basically just a data look up so I think best to use a data structure as it's a better descriptor of what we're doing!

File: src/features/domain-settings/Steps/Footers/CompleteStepFooter.tsx (hunk 2 above)
Comment (colbr) [reply src/features/domain-settings/Steps/Footers/CompleteStepFooter.tsx:36]: Nicely done

File: src/features/domain-settings/Steps/Footers/DetailsStepFooter.tsx [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+				/>
+			)}
+
+			<div className={styles.Buttons}>
+				{metadataLockedStatus && isLockedByOwner && (
+					<>
+						<IconLock1 className={styles.LockedIcon} />
+						<Button
+							onPress={() =>
+								onSubmit(ConfirmActionType.UNLOCK, steps[1], 'Unlock Metadata?')
```
Comment (colbr) [inline src/features/domain-settings/Steps/Footers/DetailsStepFooter.tsx:50]: Couple things that could make this cleaner

1. Wrap function

```js
const onSubmitDetails = (action, formHeader) => {
    return onSubmit(action, steps[1], formHeader)
}
```

2. Set more rigid responsibilities

I think the `DetailsStepFooter` should only be responsible for calling `onSubmit` with an action type, and should leave the current step and form header to the form hook. The `DetailsStepFooter` shouldn't be worried about what the current step is, or what the header text should be.

Ideally, you'd just call `onSubmit(action)`, and the form hook would set internal state for `step` and `formHeader` based on the current step and the action called. Keeping your components as dumb as possible (having as few responsibilities as possible) is a great way to simplify.

I may be missing something in the functionality though so if the above doesn't apply then all good!

File: src/features/domain-settings/hooks/useDomainSettingsData.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+	const { account } = useWeb3();
+
+	const { data: domain, isLoading: isLoadingDomainData } =
+		useDomainData(domainId);
+
+	const { data: metadata, isLoading: isLoadingMetadata } =
+		useDomainMetadata(domainId);
+
+	const { data: metadataLockedStatus, isLoading: isLoadingLockedStatus } =
```
Comment (colbr) [inline src/features/domain-settings/hooks/useDomainSettingsData.ts:21]: Are there multiple `metadataLockedStatus` values, or is it just `true`, `false`, or `undefined`? If it's the latter, can we change this to `isMetadataLocked`? I generally like to start all boolean variable names with `is` so it's obvious what values they can contain!

**Review by colbr — DISMISSED** [review]
Looking awesome @domw30. Left a couple of comments that might help you push it over the line! I'm happy with where it's at currently if you want to merge and iterate, but also happy for you to work on this one for a bit if you're wanting to get it perfect before moving on!

File: src/features/ui/Switch/Switch.tsx (hunk 6 above)
Comment (domw30) [reply src/features/ui/Switch/Switch.tsx:1]: Added to backlog :)

File: src/features/domain-settings/Steps/ConfirmStep.tsx (hunk 12 above)
Comment (domw30) [reply src/features/domain-settings/Steps/ConfirmStep.tsx:86]: Updated!

File: src/features/domain-settings/Steps/Footers/DetailsStepFooter.tsx (hunk 13 above)
Comment (domw30) [reply src/features/domain-settings/Steps/Footers/DetailsStepFooter.tsx]: This should be looking cleaner now, updated mate!

File: src/features/domain-settings/DomainSettings.constants.ts [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+		label: 'Your changes have been saved and the metadata is locked',
+	},
+	[ConfirmActionType.SAVE_WITHOUT_LOCKING]: {
+		label: 'Your changes have been saved',
+	},
+};
+
+// /**************************
+//  * CONFIRM_STEP_HEADER_TEXT
+//  **************************/
```
Comment (colbr) [inline src/features/domain-settings/DomainSettings.constants.ts:28]: nit: extra `//` before doc comment

File: src/features/domain-settings/DomainSettings.constants.ts [hunk 16]
```diff
@@ -0,0 +1,110 @@
+import { ConfirmActionType } from './DomainSettings.types';
+
+/**************************
+ * COMPLETE_STEP_LABEL_TEXT
+ **************************/
```
Comment (colbr) [inline src/features/domain-settings/DomainSettings.constants.ts:5]: nit: for doc comments, write them out in plain English. The variable name is already down below so the doc comment should just give a little description of what you're looking at. Also worth noting, if you put the doc comment above the exported function, your IDE will show you the doc comment whenever you use the function. E.g.
```
/**
 * Adds two numbers
 */
const add = ...
```
Will show "add -> Adds two numbers" when you hover over the function. Very handy 😊

File: src/features/domain-settings/DomainSettings.types.ts [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+	image_full: string;
+	previewImage: string;
+	animation_url: string;
+	stakingRequests: StakingRequests;
+	isMintable: boolean;
+	isBiddable: boolean;
+	gridViewByDefault: boolean;
+	customDomainHeader: boolean;
+	customDomainHeaderValue: string;
+}
```
Comment (colbr) [inline src/features/domain-settings/DomainSettings.types.ts:57]: There's some duplication here - would it make things simpler if you extended the `FieldValues` type? i.e.
```ts
export interface FieldValues = {
    title: string;
	description: string;
	isMintable: boolean;
	isBiddable: boolean;
	gridViewByDefault: boolean;
	customDomainHeader: boolean;
	customDomainHeaderValue: string;
}

export interface DetailsFormSubmit extends FieldValues = {
    title: string;
	description: string;
	attributes: any;
	image: string;
	image_full: string;
	previewImage: string;
	animation_url: string;
	stakingRequests: StakingRequests;
	isMintable: boolean;
	isBiddable: boolean;
	gridViewByDefault: boolean;
	customDomainHeader: boolean;
	customDomainHeaderValue: string;
}
```
Just a suggestion, we can refactor later if need be.

File: src/features/domain-settings/DomainSettingsHeader.tsx [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+const cx = classNames.bind(styles);
+
+interface DomainSettingsHeaderProps {
+	zna: string;
+	stepId: string;
+	headerText: string;
+	isTransactionLoading: boolean;
+	onClose: () => void;
+	onStepBarUpdate: (step: Step) => void;
```
Comment (colbr) [inline src/features/domain-settings/DomainSettingsHeader.tsx:19]: nit: would this function name make more sense as `onStepBarNavigation`? `onStepBarUpdate` is a little bit ambiguous.

File: src/features/domain-settings/hooks/useDomainSettingsForm.tsx [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+	};
+
+	const onCheckMetadataLockStatus = useCallback(async () => {
+		const isDomainMetadataLocked = await sdk.isDomainMetadataLocked(
+			domainId,
+			provider.getSigner(),
+		);
+
+		return isDomainMetadataLocked;
+	}, [domainId, provider]);
```
Comment (colbr) [inline src/features/domain-settings/hooks/useDomainSettingsForm.tsx:90]: nit: would be good to use `react-query` here, but we can do that as an iteration

File: src/features/domain-settings/hooks/useDomainSettingsForm.tsx [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+			action === ConfirmActionType.UNLOCK ||
+			action === ConfirmActionType.LOCK
+		) {
+			return sdk.lockDomainMetadata;
+		} else if (action === ConfirmActionType.SAVE_AND_LOCK) {
+			return sdk.setAndLockDomainMetadata;
+		} else {
+			return sdk.setDomainMetadata;
+		}
+	};
```
Comment (colbr) [inline src/features/domain-settings/hooks/useDomainSettingsForm.tsx:126]: nit: move this one outside of the hook since it doesn't rely on any state/props.

**Review by colbr — APPROVED** [review]
Couple nits but looks solid - nice!

**Review by colbr — APPROVED** [review]
